File: java/NetworkEmulationJava/src/main/java/com/github/insanusmokrassar/KotlinVSJava/JavaComparisonModule/NetworkEmulationJava/NetworkEmulation.java
package com.github.insanusmokrassar.KotlinVSJava.JavaComparisonModule.NetworkEmulationJava;

import com.github.insanusmokrassar.KotlinVSJava.report_formatter.LogPoint;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.*;

import static com.github.insanusmokrassar.KotlinVSJava.report_formatter.LogPointKt.*;

public class NetworkEmulation {
    private static final String testName = "NetworkEmulation.Java";

    public static void main(String[] args) throws ExecutionException, InterruptedException {
        int countOfRequests = 1000000;
        try {
            countOfRequests = Integer.valueOf(args[0]);
        } catch (Exception ignored) {}

        ExecutorService executor = Executors.newFixedThreadPool(8);
        Network network = new NetworkImpl();

        List<LogPoint> points = new ArrayList<>();
        points.add(new LogPoint(testName, START_TEST));

        List<Request> requests = new ArrayList<>();

        for (int i = 0; i < countOfRequests; i++) {
            requests.add(
                new Request(i, network)
            );
        }
        points.add(new LogPoint(testName, REQUESTS_FILLED));

        List<Future<Integer>> futures = new ArrayList<>();

        for (final Request request : requests) {
            Future<Integer> future = executor.submit(
                    new Callable<Integer>() {
                        @Override
                        public Integer call() throws Exception {
                            return request.request();
                        }
                    }
            );

            futures.add(future);
        }
        points.add(new LogPoint(testName, ALL_TASKS_CREATED));

        for (Future<Integer> future : futures) {
            future.get();
        }
        points.add(new LogPoint(testName, ALL_TASKS_COMPLETED));

        points.add(new LogPoint(testName, COMPLETE_TEST));
    }
}

File: KotlinComparisonModule/NetworkEmulationKotlin/src/main/kotlin/com/github/insanusmokrassar/KotlinVSJava/KotlinComparisonModule/NetworkEmulationKotlin/NetworkEmulation.kt
package com.github.insanusmokrassar.KotlinVSJava.KotlinComparisonModule.NetworkEmulationKotlin

import com.github.insanusmokrassar.KotlinVSJava.report_formatter.*
import kotlinx.coroutines.*
import java.util.concurrent.Executors

private const val testName = "NetworkEmulation.Kotlin"

fun main(args: Array<String>) {
    val countOfRequests = args.firstOrNull() ?.toIntOrNull() ?: 1000000
    val dispatcher = Executors.newFixedThreadPool(8).asCoroutineDispatcher()

    val network: Network = NetworkImpl()

    val points = arrayListOf<LogPoint>()
    points.add(LogPoint(testName, START_TEST))

    val requests = (0 until countOfRequests).map {
        Request(it, network)
    }
    points.add(LogPoint(testName, REQUESTS_FILLED))

    runBlocking(dispatcher) {
        requests.map {
            async {
                it.request()
            }
        }.also {
            points.add(LogPoint(testName, ALL_TASKS_CREATED))
        }.forEach {
            it.await()
        }
        points.add(LogPoint(testName, ALL_TASKS_COMPLETED))
    }
    points.add(LogPoint(testName, COMPLETE_TEST))

    dispatcher.close()
    printPoints(points)
}
